clang/docs/SYCLSupport.rst

```diff
@@ -1,119 +1,542 @@
+require special handling, and because C++17 does not support function templates
+with multiple function parameter packs, an idiom is used to pass the kernel
+arguments and special subobjects separately.
+This idiom is best explained by way of an example.
+
+Consider the earlier example of a call to ``single_task()`` that passes a
+lambda that captures variables of type ``int`` and ``std::stream``.
+``std::stream`` is an example of a SYCL type that requires special handling in
+kernel arguments.
+The call to ``kernel_entry_point<KernelName>(kernelFunc)`` in the implementation
+of ``single_task()`` results in an implicit call to ``sycl_kernel_launch`` that
+looks similar to the following (the access to the captured copy of ``sout`` via
+``kernelFunc.sout`` is not valid C++ syntax, but the compiler can generate such
+accesses).
+
+.. code-block:: C++
+
+   sycl_kernel_launch<KernelName>("kernel-entry-point", kernelFunc)(kernelFunc.sout)
+
+The SYCL kernel name type, ``KernelName``, is passed as an explicit template
+type argument for convenient use by the SYCL run-time library if desired.
+The first function argument is the name of the offload kernel entry point
+function generated for the SYCL kernel denoted by ``KernelName``.
+This argument names the function that the SYCL run-time library, in conjunction
+with an offload backend, needs to resolve and execute on the selected device
+in order to execute the SYCL kernel.
+The remaining arguments, for which there is just one in this case,
+``kernelFunc``, are the kernel arguments.
+Kernel arguments that contain a special subobject are passed as lvalues; those
+that don't are passed as xvalues (and may therefore be moved from though such
+types are unlikely to have optimized move operations since they are required to
+be bit-copyable types).
+
+The result of the ``sycl_kernel_launch`` call must be a function object (e.g.,
+the result of a lambda expression, potentially one that captures references to
+the kernel arguments).
+The special subobjects are then passed as lvalues in an invocation of the
+resulting function object.
+This is the idiom mentioned earlier; this effectively allows for a call with
+two function parameter packs.
+A SYCL run-time library is expected to define the ``sycl_kernel_launch``
+template in a form similar to this:
+
+.. code-block:: C++
+
+   namespace sycl {
+     class handler {
+       template <typename KernelName, typename... KernelArgs>
+       auto sycl_kernel_launch(const char *entryPointName, KernelArgs &... args) {
+         return [&] (auto &... subobjects) {
+           // Process the kernel arguments and special subobjects, schedule
+           // execution of 'entryPointName' on the selected device, and
+           // return a type-list object sepcifying additional parameters to
+           // add to the offload kernel entry point function (see below).
+           return detail::type_list<...>{};
+         };
+       }
+       ...
+     };
+  }
```

Comment:
IMO, I'd restructure to include this code snippet early and then describe what it does. I was able to follow the description above only because I was already familiar with it.

Reply:
Start with the full snippet and, maybe, repeat parts of it when talking about something specific.